clients/da-vinci-client/src/main/java/com/linkedin/davinci/blob/BlobSnapshotManager.java

```diff
@@ -0,0 +1,124 @@
+
+  /**
+   * Constructor for the BlobSnapshotManager
+   */
+  public BlobSnapshotManager(String basePath, long snapshotRetentionTime) {
+    this.basePath = basePath;
+    this.snapshotRetentionTime = snapshotRetentionTime;
+    this.snapShotTimestamps = new HashMap<>();
+    this.concurrentSnapshotUsers = new HashMap<>();
+  }
+
+  /**
+   * Checks if the current snapshot of the partition is stale
+   */
+  private boolean isSnapshotStale(String topicName, int partitionId) {
+    if (!snapShotTimestamps.containsKey(topicName) || !snapShotTimestamps.get(topicName).containsKey(partitionId)) {
+      return true;
+    }
+    return System.currentTimeMillis() - snapShotTimestamps.get(topicName).get(partitionId) > snapshotRetentionTime;
+  }
+
+  /**
+   * Updates the snapshot of the hybrid store
+   */
+  private void updateHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    if (fullPathForPartitionDBSnapshot.isEmpty()) {
+      LOGGER.error(
+          "Error creating snapshot directory for topic: {} and partition: {} due to not path existing",
+          topicName,
+          partitionId);
+      return;
+    }
+    try {
+      Checkpoint checkpoint = createCheckpoint(rocksDB);
+
+      LOGGER.info("Creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+      checkpoint.createCheckpoint(fullPathForPartitionDBSnapshot);
+
+      LOGGER.info("Finished creating snapshots in directory: {}", fullPathForPartitionDBSnapshot);
+    } catch (RocksDBException e) {
+      throw new VeniceException(
+          "Received exception during RocksDB's snapshot creation in directory " + fullPathForPartitionDBSnapshot,
+          e);
+    }
+  }
+
+  /**
+   * Checks if the snapshot is stale, if it is and no one is using it, it updates the snapshot,
+   * otherwise it increases the count of people using the snapshot
+   */
+  public void getHybridSnapshot(RocksDB rocksDB, String topicName, int partitionId) {
+    if (rocksDB == null || topicName == null) {
+      throw new IllegalArgumentException("RocksDB instance and topicName cannot be null");
+    }
+    String fullPathForPartitionDBSnapshot = RocksDBUtils.composeSnapshotDir(this.basePath, topicName, partitionId);
+    if (!snapShotTimestamps.containsKey(topicName)) {
+      snapShotTimestamps.put(topicName, new HashMap<>());
+      concurrentSnapshotUsers.put(topicName, new HashMap<>());
+    }
+    if (isSnapshotStale(topicName, partitionId)) {
+      if (concurrentSnapshotUsers.get(topicName).get(partitionId) == null
+          || concurrentSnapshotUsers.get(topicName).get(partitionId).get() == 0) {
+        updateHybridSnapshot(rocksDB, topicName, partitionId);
+        snapShotTimestamps.get(topicName).put(partitionId, System.currentTimeMillis());
+      }
+      if (getConcurrentUsers(rocksDB, topicName, partitionId) == 0) {
```

Comment:
I found there is an issue around this: There is no way we decrease the counter of the snapshot transfer user. We need to think about this. Essentially, BlobTransferManger knows, maybe can we add another API called by BlobTransferManger to invoke to decrease the counter by 1?

Reply:
Implemented the functionality to decrease counter, just not sure how to check if someone is no longer using the snapshot to apply it

Reply:
As of now, we should not be worrying about whether someone is no longer using this. The caller: BlobTransferManger, is supposed to take care of this and invoke this method when the transfer is completed. We just need to make sure snapshot creation is thread safe.

Reply:
By saying 

> BlobTransferManger, is supposed to take care of this

I mean, when we are integrating hybrid snapshot logic into BlobTransferManager, we should make sure it is invoked properly.